libphobos/src/std/math.d

```diff
@@ -4456,61 +4470,58 @@ private:
 private:
     static uint getIeeeFlags()
     {
-        version(D_InlineAsm_X86)
-        {
-            asm pure nothrow @nogc
-            {
-                 fstsw AX;
-                 // NOTE: If compiler supports SSE2, need to OR the result with
-                 // the SSE2 status register.
-                 // Clear all irrelevant bits
-                 and EAX, 0x03D;
-            }
-        }
-        else version(D_InlineAsm_X86_64)
+        version(InlineAsm_X86_Any)
         {
-            asm pure nothrow @nogc
+            ushort sw;
+            asm pure nothrow @nogc { fstsw sw; }
+
+            // OR the result with the SSE2 status register (MXCSR).
+            if (haveSSE)
             {
-                 fstsw AX;
-                 // NOTE: If compiler supports SSE2, need to OR the result with
-                 // the SSE2 status register.
-                 // Clear all irrelevant bits
-                 and RAX, 0x03D;
+                uint mxcsr;
+                asm pure nothrow @nogc { stmxcsr mxcsr; }
+                return (sw | mxcsr) & EXCEPTIONS_MASK;
             }
+            else return sw & EXCEPTIONS_MASK;
         }
         else version (GNU)
         {
-            uint result;
-            version (X86)
+            version (X86_Any)
             {
+                ushort sw;
                 asm pure nothrow @nogc
                 {
-                    "fstsw %%ax; andl $0x03D, %%eax;" : "=a" result;
+                    "fstsw %0" : "=a" (sw);
                 }
-            }
-            else version (X86_64)
-            {
-                asm pure nothrow @nogc
+                // OR the result with the SSE2 status register (MXCSR).
+                if (haveSSE)
                 {
-                    "fstsw %%ax; andq $0x03D, %%rax;" : "=a" result;
+                    uint mxcsr;
+                    asm pure nothrow @nogc
+                    {
+                        "stmxcsr %0" : "=m" (mxcsr);
```

Comment:
Yes. I don't doubt this works fine at runtime but on ARM the assembler simply refuses to even assemble FPU instructions without passing certain FPU target flags to the assembler. For i386 the documentation says:
>-march=CPU[+EXTENSION…]
>
> This option specifies the target processor. **The assembler will issue an error message if an attempt is made to assemble an instruction which will not execute on the target processor.** The following processor names are recognized: i8086, i186, i286, i386, i486, i586, i686, pentium, pentiumpro, pentiumii, pentiumiii, pentium4, prescott, nocona, core, core2, corei7, l1om, k1om, iamcu, k6, k6_2, athlon, opteron, k8, amdfam10, bdver1, bdver2, bdver3, bdver4, znver1, btver1, btver2, generic32 and generic64.
>
>In addition to the basic instruction set, the assembler can be told to accept various extension mnemonics. For example, -march=i686+sse4+vmx extends i686 with sse4 and vmx. The following extensions are currently supported: 8087, 287, 387, 687, no87, no287, no387, no687, mmx, nommx, sse, sse2, sse3, ssse3, sse4.1, sse4.2, sse4, nosse, nosse2, nosse3, nossse3, nosse4.1, nosse4.2, nosse4, avx, avx2, noavx, noavx2, adx, rdseed, prfchw, smap, mpx, sha, rdpid, ptwrite, cet, prefetchwt1, clflushopt, se1, clwb, avx512f, avx512cd, avx512er, avx512pf, avx512vl, avx512bw, avx512dq, avx512ifma, avx512vbmi, avx512_4fmaps, avx512_4vnniw, avx512_vpopcntdq, noavx512f, noavx512cd, noavx512er, noavx512pf, noavx512vl, noavx512bw, noavx512dq, noavx512ifma, noavx512vbmi, noavx512_4fmaps, noavx512_4vnniw, noavx512_vpopcntdq, vmx, vmfunc, smx, xsave, xsaveopt, xsavec, xsaves, aes, pclmul, fsgsbase, rdrnd, f16c, bmi2, fma, movbe, ept, lzcnt, hle, rtm, invpcid, clflush, mwaitx, clzero, lwp, fma4, xop, cx16, syscall, rdtscp, 3dnow, 3dnowa, sse4a, sse5, svme, abm and padlock. Note that rather than extending a basic instruction set, the extension mnemonics starting with no revoke the respective functionality.
>
>When the .arch directive is used with -march, the .arch directive will take precedent. 
https://sourceware.org/binutils/docs/as/i386_002dOptions.html#i386_002dOptions

I'm just wondering how this works in practice. If we configure GCC with an SSE target we know  SSE is available statically anyway. If we want to do a runtime check we have to compile for non-SSE target so GCC does not emit SSE instructions in codegen, but then `as` will refuse to assemble these instructions, as far as I understand. Do we somehow have to emit the `.arch` directive?